src/math/src/number.rs

```diff
@@ -1,18 +1,523 @@
+use crate::Result;
+use std::cmp::Ordering;
+
+#[derive(Default, Debug, Clone, Copy)]
 /// Represent a number
-pub struct Number(f64);
+pub struct Number {
+    inner: f64,
+}
+
+#[derive(Default, Debug, Clone, Copy)]
+#[non_exhaustive]
+/// Radix to use to represent a Number
+pub enum Radix {
+    /// Binary
+    Bin,
+    /// Octal
+    Oct,
+    #[default]
+    /// Decimal - default
+    Dec,
+    /// Hexadecimal
+    Hex,
+}
+
+impl<T: Into<f64>> From<T> for Number {
+    fn from(v: T) -> Self {
+        Self { inner: v.into() }
+    }
+}
 
 impl AsRef<f64> for Number {
     fn as_ref(&self) -> &f64 {
-        &self.0
+        &self.inner
     }
 }
 
 impl Number {
     /// 0.0
-    pub const ZERO: Number = Self(0.0);
+    pub const ZERO: Self = Self { inner: 0.0 };
+
+    /// 1.0
+    pub const ONE: Self = Self { inner: 1.0 };
+
+    /// 3.14159...
+    pub const PI: Self = Self {
+        inner: std::f64::consts::PI,
+    };
 
+    /// 2.71828...
+    pub const E: Self = Self {
+        inner: std::f64::consts::E,
+    };
+
+    /// Get the formatted string of a number
+    ///
+    /// ```
+    /// # use math::number::{Number, Radix};
+    /// let pi = Numper::PI;
+    /// let zero = Number::ZERO;
+    /// let neg = Number::from(-0.1);
     ///
-    pub fn to_string_radix(&self, _radix: u8) -> String {
+    /// let precision = 6;
+    ///
+    /// assert_eq!(zero.to_string(Radix::Bin, precision), "0");
+    /// assert_eq!(zero.to_string(Radix::Oct, precision), "0");
+    /// assert_eq!(zero.to_string(Radix::Dec, precision), "0");
+    /// assert_eq!(zero.to_string(Radix::Hex, precision), "0");
+    ///
+    /// assert_eq!(pi.to_string(Default::default(), precision), "3.141592");
+    /// assert_eq!(pi.to_string(Radix::Bin, precision), "11.001001");
+    /// assert_eq!(pi.to_string(Radix::Oct, precision), "3.110375");
+    /// assert_eq!(pi.to_string(Radix::Hex, precision), "3.243F6A");
+    ///
+    /// assert_eq!(neg.to_string(Default::default(), precision), "-0.1");
+    /// assert_eq!(neg.to_string(Radix::Bin, precision), "-0.00011");
+    /// assert_eq!(neg.to_string(Radix::Oct, precision), "-0.063146");
+    /// assert_eq!(neg.to_string(Radix::Hex, precision), "-0.199999");
+    /// ```
+    pub fn to_string(&self, radix: Radix, precision: u8) -> String {
         todo!()
     }
 }
+
+impl Number {
+    /// Generate a random number in range of <0, 1>
+    ///
+    /// ```
+    /// assert_ne!(Number::random(), Number::random());
+    /// assert_ge!(Number::random(), Number::Zero);
+    /// assert_le!(Number::random(), Number::One);
+    /// ```
+    pub fn random() -> Self {
+        todo!()
+    }
+
+    /// Add two numbers together
+    ///
+    /// ```
+    /// # use math::Number;
+    ///
+    /// let a = Number::from(0.1);
+    /// let b = Number::from(0.2);
+    ///
+    /// assert_eq!(a.add(b), Ok(Number::from(0.3)));
```

Comment:
It is possible that these will fail, because:
```rust
fn main() {
    let a = 0.1 + 0.2;      // Equals to 0.30000000000000004
    let b = 0.3;            // Equals to 0.3
    if a == b {
        println!("Equal");
    } else {
        println!("NOT equal");      // Will print this
    }
}
```
I have noticed, that number `0.1` is used in other unit tests as well. Maybe we should use numbers such as `0.5` and `0.25`, which are finite and short in binary.

Reply:
This is actually what I want, because for those float point errors in binary representation shouldn't be in a calculator
I'm currently looking for other way to store a number instead of `f64` and that test must pass.

Reply:
Then we could use [bigdecimal](https://docs.rs/bigdecimal/latest/bigdecimal) crate for this

Reply:
That one has problem with performance since the number is always allocated on the heap, also it doesn't have pow function...
I'm planning to go with the [`Ratio`](https://docs.rs/num/latest/num/rational/struct.Ratio.html) type because it fits everything we need.

Reply:
Alright

Reply:
However, won't there be a problem representing irrational numbers and (for example) trigonometric functions?